File: android/DigitRecognizer/app/src/main/java/com/mzm/sample/digit_recognizer/CustomView.java
package com.mzm.sample.digit_recognizer;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.MotionEvent;
import android.view.View;

/**
 * A custom view that allows the user to draw the digits
 * Note we are using a black background for the canvas and white stroke
 * to better match the MNIST training data
 */
public class CustomView extends View {

    private Canvas canvas;
    private Paint paint;
    private Path path;
    private Bitmap bitmap;
    private Context context;

    private float x, y;
    private static final float TOUCH_TOLERANCE = 4;

    public CustomView(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        init();
    }

    private void init() {
        DisplayMetrics dm = context.getResources().getDisplayMetrics();
        bitmap= Bitmap.createBitmap(dm.widthPixels, dm.heightPixels, Bitmap.Config.ARGB_8888);
        canvas =  new Canvas(bitmap);
        path = new Path();
        paint = new Paint();

        paint.setAntiAlias(true);       // smooth out edges of drawing
        paint.setDither(true);
        paint.setColor(Color.WHITE);    // set stroke color to black
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeWidth(48f);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawColor(Color.BLACK);  // set canvas background to white
        canvas.drawBitmap(bitmap, 0, 0, paint);
        canvas.drawPath(path, paint);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        bitmap = bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
        canvas = new Canvas(bitmap);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x = event.getX();
        float y = event.getY();

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                touch_start(x, y);
                invalidate();
                break;
            case MotionEvent.ACTION_MOVE:
                touch_move(x, y);
                invalidate();
                break;
            case MotionEvent.ACTION_UP:
                touch_up();
                invalidate();
                break;
        }
        return true;
    }

    private void touch_start(float x, float y) {
        path.reset();
        path.moveTo(x, y);
        this.x = x;
        this.y = y;
    }
    private void touch_move(float x, float y) {
        float dx = Math.abs(x - this.x);
        float dy = Math.abs(y - this.y);
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            path.quadTo(this.x, this.y, (x + this.x)/2, (y + this.y)/2);
            this.x = x;
            this.y = y;
        }
    }
    private void touch_up() {
        path.lineTo(x, y);
        canvas.drawPath(path, paint);
        path.reset();
    }

    public void clear() {
        path.reset();
        bitmap = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), Bitmap.Config.ARGB_8888);
        canvas = new Canvas(bitmap);
        invalidate();
    }

    public Bitmap getBitmap(int width, int height) {
        return Bitmap.createScaledBitmap(bitmap, 28, 28, false);
    }
}

File: android/kotlin/DigitRecognizer/app/src/main/java/com/mzm/sample/digit_recognizer/CustomView.kt
package com.mzm.sample.digit_recognizer

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

/**
 * A custom view that allows the user to draw the digits
 * Note we are using a black background for the canvas and white stroke
 * to better match the MNIST training data
 */
class CustomView(context: Context, attrs: AttributeSet) : View(context, attrs) {

    private lateinit var canvas: Canvas
    private lateinit var paint: Paint
    private lateinit var path: Path
    private lateinit var bitmap: Bitmap

    private var currX = 0f
    private var currY = 0f

    init {
        init()
    }

    private fun init() {
        val dm = context.resources.displayMetrics
        bitmap = Bitmap.createBitmap(dm.widthPixels, dm.heightPixels, Bitmap.Config.ARGB_8888)
        canvas = Canvas(bitmap!!)
        path = Path()
        paint = Paint()

        paint.isAntiAlias = true       // smooth out edges of drawing
        paint.isDither = true
        paint.color = Color.WHITE    // set stroke color to black
        paint.style = Paint.Style.STROKE
        paint.strokeJoin = Paint.Join.ROUND
        paint.strokeWidth = 48f
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawColor(Color.BLACK)  // set canvas background to black
        canvas.drawBitmap(bitmap!!, 0f, 0f, paint)
        canvas.drawPath(path!!, paint!!)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        canvas = Canvas(bitmap)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x = event.x
        val y = event.y

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                touch_start(x, y)
                invalidate()
            }
            MotionEvent.ACTION_MOVE -> {
                touch_move(x, y)
                invalidate()
            }
            MotionEvent.ACTION_UP -> {
                touch_up()
                invalidate()
            }
        }
        super.onTouchEvent(event)
        return true
    }

    private fun touch_start(x: Float, y: Float) {
        path.reset()
        path.moveTo(x, y)
        this.currX = x
        this.currY = y
    }

    private fun touch_move(x: Float, y: Float) {
        val dx = Math.abs(x - this.currX)
        val dy = Math.abs(y - this.currY)
        if (dx >= TOUCH_TOLERANCE || dy >= TOUCH_TOLERANCE) {
            path.quadTo(this.currX, this.currY, (x + this.currX) / 2, (y + this.currY) / 2)
            this.currX = x
            this.currY = y
        }
    }

    private fun touch_up() {
        path.lineTo(x, y)
        canvas.drawPath(path, paint)
        path.reset()
    }

    fun clear() {
        path.reset()
        bitmap = Bitmap.createBitmap(bitmap.width, bitmap.height, Bitmap.Config.ARGB_8888)
        canvas = Canvas(bitmap)
        invalidate()
    }

    fun getBitmap(): Bitmap {
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawColor(Color.BLACK)
        draw(canvas)
        return bitmap
    }

    companion object {
        private val TOUCH_TOLERANCE = 4f
    }
}
